rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -1640,6 +1731,83 @@ const rosbag2_transport::PlayOptions & PlayerImpl::get_play_options()
   return play_options_;
 }
 
+void PlayerImpl::print_progress_bar_help_str() const
+{
+  std::string help_str;
+  if (enable_progress_bar_) {
+    if (progress_bar_update_always_) {
+      help_str = "Progress bar enabled for every message.";
+    } else {
+      std::ostringstream oss;
+      oss << "Progress bar enabled at " <<
+        play_options_.progress_bar_print_frequency << " Hz";
+      help_str = oss.str();
+    }
+    RCLCPP_INFO_STREAM(owner_->get_logger(), help_str);
+    std::string help_str2 = "Progress bar t: time, d: duration";
+    RCLCPP_INFO_STREAM(owner_->get_logger(), help_str2);
+    std::string help_str3 = "Progress bar [?]: [R]unning, [P]aused, [B]urst, [D]elayed";
+    RCLCPP_INFO_STREAM(owner_->get_logger(), help_str3);
+  } else {
+    help_str = "Progress bar disabled";
+    RCLCPP_INFO_STREAM(owner_->get_logger(), help_str);
+  }
+}
+
+void PlayerImpl::update_progress_bar_check_frequency(
+  const rcutils_time_point_value_t & timestamp,
+  const PlayerStatus & status)
+{
+  if (!enable_progress_bar_) {
+    return;
+  }
+
+  // If we are not updating the progress bar for every call, check if we should update it now
+  // based on the frequency set by the user
+  if (!progress_bar_update_always_) {
+    std::chrono::steady_clock::time_point steady_time_now = std::chrono::steady_clock::now();
+    if (std::chrono::duration_cast<std::chrono::nanoseconds>(
+      steady_time_now - progress_bar_last_time_updated_).count() < progress_bar_update_period_)
+    {
+      return;
+    }
+    progress_bar_last_time_updated_ = steady_time_now;
+  }
+
+  cout_progress_bar(timestamp, status);
+}
+
+void PlayerImpl::update_progress_bar(const PlayerStatus & status) const
+{
+  if (!enable_progress_bar_) {
+    return;
+  }
+
+  // Update progress bar irrespective of the frequency set by the user.
+  // Overwrite the input status if we are in a special case.
+  if (is_delayed_status_) {
+    cout_progress_bar(starting_time_, PlayerStatus::DELAYED);
+  } else {
+    cout_progress_bar(std::min(clock_->now(), ending_time_), status);
+  }
+}
+
+void PlayerImpl::cout_progress_bar(
+  const rcutils_time_point_value_t & timestamp,
+  const PlayerStatus & status) const
+{
+  const double current_time_secs = RCUTILS_NS_TO_S(static_cast<double>(timestamp));
+  const double progress_secs = current_time_secs - starting_time_secs_;
+  // Print progress bar ending with carriage return '/r' so it will be overwritten by next print
+  // Spaces before '\r' are to clear any previous progress bar in case the new one is shorter,
+  // which can happen when the playback starts a new loop.
+  std::ostringstream oss;
+  oss << " Bag t: " << std::setw(13) << std::fixed << std::setprecision(9) <<
+    current_time_secs << " d: " << std::setprecision(9) << progress_secs <<
+    " [" << static_cast<char>(status) << "]      \r";
```

Comment:
Having this many digits makes it hard to easily and quickly read the progress. Compare

```
4.194342136/16.974378413
```

with

```
4.19/16.97
```

The user also doesn't really need this amount of precision, right? In ROS 1, rosbag used 6 digits (IIUC), which is still less than the 9 digits used here, but I think it's still more than what's really needed.

Something to consider.

Reply:
I agree. For the duration, we don't need so many digits after the floating point. IMO, 3 digits would be enough, up to millisecond accuracy.

cc: @nicolaloi

Reply:
Update. After more consideration, I think even 2 digits will be enough. I can't imagine a use case when higher precision would be usefull for the information about duration.

Reply:
Maybe there was a miscommunication yesterday, after https://github.com/ros2/rosbag2/pull/1836#issuecomment-2519646020 I changed the bag duration from milliseconds to nanoseconds, but probably the comment was referring to the bag timestamp only and not the actual bag duration.

Ok so I'll keep the timestamp with nanosecond precision, and change the bag duration to centiseconds (2 decimal digits) precision.